### PR TITLE
feat(toggle): add support for changing toggle width and height

### LINKED ISSUE #19868: bug: On ion-toggle, toggle-inner does not adapt to height and width
# Bug Report

**Ionic version:**

[x] **4.10**

**Current behavior:**

We can change height and width of a ion-toggle but toggle-inner does not adapt the respective height and width

Here is the styling that I am using:
```
<ion-toggle #onOffToggle mode='ios'></ion-toggle>
ion-toggle {
		--background: var(--color-red);
		--background-checked: var(--color-green);
		height: 20px;
		width: 50px;
	}
```
image

**Expected behavior:**

Inner circle should adapt to the respective height

Also by some way can I make toogle look like this?
image

**Steps to reproduce:**

<ion-toggle #onOffToggle mode='ios'></ion-toggle>

**Other information:**

**Ionic info:**

```
Ionic:

   Ionic CLI                     : 5.4.2
   Ionic Framework               : @ionic/angular 4.10.0
   @angular-devkit/build-angular : 0.801.3
   @angular-devkit/schematics    : 8.1.3
   @angular/cli                  : 8.1.3
   @ionic/angular-toolkit        : 2.0.0

Cordova:

   Cordova CLI       : 9.0.0 (cordova-lib@9.0.1)
   Cordova Platforms : android 8.1.0
   Cordova Plugins   : cordova-plugin-ionic-keyboard 2.2.0, cordova-plugin-ionic-webview 4.1.3, (and 5 other plugins)

Utility:

   cordova-res : not installed
   native-run  : 0.2.9

System:

   Android SDK Tools : 26.1.1 (C:\Program Files\Android\android-sdk)
   NodeJS            : v12.13.0 (C:\Program Files\nodejs\node.exe)
   npm               : 6.12.0
   OS                : Windows 10

```

### LINKED ISSUE #20474: feat: Could you provide extra css variable properties for ion-toggle in ionic 5?
In the below **.toggle-inner**  is not able to update because of shadow component.
`.toggle-inner {
    left: 2px;
    top: 2px;
    position: absolute;
    width: 28px;
    height: 28px;
    box-shadow: rgba(0, 0, 0, 0.16) 0px 3px 12px, rgba(0, 0, 0, 0.1) 0px 3px 1px;
    will-change: transform;
    contain: strict;
    transition: transform 0.3s ease 0s, width 0.12s ease-in-out 80ms, left 0.11s ease-in-out 80ms, right 0.11s ease-in-out 80ms, -webkit-transform 0.3s ease 0s;
}`
Please provide the following css variables **left, top, width, height and box-shadow** for more flexible customisation.

Thank you

### PR DESCRIPTION
**Pull request by brandyscarney**
# Toggle Customization Improvements

- Uses the proposal here with adjustments: https://github.com/ionic-team/ionic/issues/19868#issuecomment-595861808
- Updates Material Design toggle background & box-shadow to match spec
- Adds the following variables: `--handle-box-shadow`, `--handle-height`, `--handle-max-height`, `--handle-transition`, `--handle-width`, `--handle-spacing`
- Closes #19868, closes #20474

## Customizing Background

The background of the toggle track and handle can be customized using CSS variables. There are also variables for setting the background differently when the toggle is checked.

```css
ion-toggle {
  --background: #000;
  --background-checked: #7a49a5;

  --handle-background: #7a49a5;
  --handle-background-checked: #000;
}
```

Screen Shot 2020-04-29 at 3 53 24 PM

Because these variables set the `background` property, which is a shorthand, it can accept any value that the [background property](https://developer.mozilla.org/en-US/docs/Web/CSS/background) accepts.

A more complex case may involve adding an image to the handle background.

```css
ion-toggle {
  --handle-background-checked: #fff url(/assets/power-icon.png) no-repeat center / contain;
}
```

Screen Shot 2020-04-29 at 3 55 16 PM

Taking it a step further, we could use the `::before` or `::after` pseudo-elements to position text on top of the background.

```css
ion-toggle::before {
  position: absolute;

  top: 16px;
  left: 10px;

  content: "ON";

  color: white;
  font-size: 8px;

  z-index: 1;
}
```

Screen Shot 2020-04-29 at 3 25 05 PM

## Customizing Width

Adjusting the width of the toggle **smaller** will result in a narrower track, with the handle remaining the default width. If desired, set `--handle-width` to make the handle narrower.

```css
ion-toggle {
  width: 40px;
}
```

Screen Shot 2020-04-27 at 5 49 33 PM

Adjusting the width of the toggle **larger** will result in a wider track, with the handle remaining the default width. If desired, set `--handle-width` to make the handle wider.

```css
ion-toggle {
  width: 100px;
}
```

Screen Shot 2020-04-27 at 5 53 18 PM

## Customizing Height

Adjusting the height of the toggle **smaller** than the default will result in the handle height auto-sizing itself to the track. In `ios` the handle is shorter than the track (`100% - 4px`) and in `md` the handle is taller than the track (`100% + 6px`).

```css
ion-toggle {
  height: 15px;
}
```

> Note: this does not affect the handle width, width should be set using `--handle-width`.

Screen Shot 2020-04-27 at 5 54 53 PM

> Note: this does not affect the width, width should be set smaller if desired using `--handle-width`

Adjusting the height of the toggle **larger** will keep the handle in the center at the default height. This can be modified by setting `--handle-height` which can be set to any amount but will not exceed the `--handle-max-height`.

```css
ion-toggle {
  height: 60px;
}
```

> Note: this does not affect the handle width, width should be set using `--handle-width`.

Screen Shot 2020-04-27 at 5 55 23 PM

> Note: this does not affect the width, width should be set smaller or larger if desired using `--handle-width`

## Customizing Spacing

The spacing refers to the horizontal gap between the handle and the track. By default, the handle has `2px` of spacing around it in `ios` **only**. In `md` mode there is no default spacing.

To remove the **horizontal** spacing, set `--handle-spacing` to `0px`.

```css
ion-toggle {
  --handle-spacing: 0px;
}
```

Screen Shot 2020-04-28 at 6 02 10 PM

Due to the handle having a fixed height, to remove the spacing on the top and bottom, set the height to 100%.

```css
ion-toggle {
  --handle-spacing: 0px;
  --handle-height: 100%;
}
```

Screen Shot 2020-04-28 at 6 03 36 PM

## Customizing Box Shadow

The `--handle-box-shadow` can be used to change the `box-shadow` on the handle.

```css
ion-toggle {
  --handle-box-shadow: 4px 0 2px 0 red;
}
```

To target the box shadow only when the toggle is checked, target `.toggle-checked`:

```css
ion-toggle.toggle-checked {
  --handle-box-shadow: -4px 0 2px 0 red;
}
```

See the section on [customizing overflow](#customizing-overflow) to allow the `box-shadow` to overflow the toggle container.

## Customizing Border Radius

The `--handle-border-radius` can be used to change the `border-radius` on the handle.

```css
ion-toggle {
  --handle-border-radius: 14px 4px 4px 14px;
}
```

To target the `border-radius` only when the toggle is checked, target `.toggle-checked`:

```css
ion-toggle.toggle-checked {
  --handle-border-radius: 4px 14px 14px 4px;
}
```

## Customizing Overflow

Setting `overflow` on the toggle will be inherited by the toggle handle. By default, overflow is set to `hidden` in `ios` only.

The below image shows two toggles with the same box shadow but the left has the default overflow, while the right toggle has set overflow to visible:

```css
ion-toggle {
  --handle-box-shadow: 0 3px 12px rgba(255, 0, 0, 0.6), 0 3px 1px rgba(50, 70, 255, 0.6);

  overflow: visible;
}
```

Screen Shot 2020-04-27 at 6 47 28 PM

The `box-shadow` will still appear cut off due to the `contain` css property. Set this to `none` in order to overflow the toggle container:

```css
ion-toggle {
  --handle-box-shadow: 0 3px 12px rgba(255, 0, 0, 0.6), 0 3px 1px rgba(50, 70, 255, 0.6);

  overflow: visible;
  
  contain: none;
}
```

Screen Shot 2020-04-29 at 12 11 42 PM

# Screenshot Diff Preview

Using test `src/components/toggle/test/sizes`

| `master` | `branch` |
| ----------| ----------|
| ios-master | ios-branch |
| md-master | md-branch |

> Note: this test doesn't exist on master, but copy/pasting it to master is how I am able to get the screenshots

### CODE REVIEW

**Comment by brandyscarney** [comment]
This is ready for review now, dev release `5.1.0-dev.202004282251.78e0507`

File: core/src/components/toggle/toggle.md.scss [hunk 1]
```diff
@@ -33,73 +36,28 @@
...
-}
-
-// Material Design Toggle Inner Knob: Checked
-// -----------------------------------------
-
-:host(.toggle-checked) .toggle-inner {
-  @include transform(translate3d($toggle-md-track-width - $toggle-md-handle-width, 0, 0));
-
-  background: var(--handle-background-checked);
+  will-change: background-color;
```
Comment (liamdebeasi) [inline core/src/components/toggle/toggle.md.scss:50]:
```suggestion
  will-change: background-color, transform;
```

Looks like the transform changes on toggle-inner too: https://github.com/ionic-team/ionic/pull/21050/files#diff-06267e2174a1a50220f0594452a3d09cR130

**Comment by liamdebeasi** [comment]
Also can we update the usage docs with more examples on how to customize the toggle? There's a lot of good information in your explainer in the initial PR comment that could be helpful to other devs when they are looking through the docs.

**Comment by brandyscarney** [comment]
Yeah I was wondering how we should add the customization documentation. I definitely agree it needs to be added somewhere. Let me finish up the outstanding questions and I'll work on that.

**Pull request merged**